src/mlpack/tests/q_learning_test.cpp

```diff
@@ -0,0 +1,133 @@
+/**
+ * @file q_learning_test.hpp
+ * @author Shangtong Zhang
+ *
+ * Test for Q-Learning implementation
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#include <mlpack/core.hpp>
+
+#include <mlpack/core/optimizers/adam/adam.hpp>
+#include <mlpack/core/optimizers/sgd/sgd.hpp>
+#include <mlpack/core/optimizers/gradient_descent/gradient_descent.hpp>
+#include <mlpack/methods/ann/ffn.hpp>
+#include <mlpack/methods/ann/init_rules/orthogonal_init.hpp>
+#include <mlpack/methods/ann/init_rules/gaussian_init.hpp>
+#include <mlpack/methods/ann/layer/layer.hpp>
+#include <mlpack/methods/reinforcement_learning/q_learning.hpp>
+#include <mlpack/methods/reinforcement_learning/environment/mountain_car.hpp>
+#include <mlpack/methods/reinforcement_learning/environment/cart_pole.hpp>
+#include <mlpack/methods/reinforcement_learning/policy/greedy_policy.hpp>
+
+#include <boost/test/unit_test.hpp>
+#include "test_tools.hpp"
+
+using namespace mlpack;
+using namespace mlpack::ann;
+using namespace mlpack::optimization;
+using namespace mlpack::rl;
+
+BOOST_AUTO_TEST_SUITE(QLearningTest);
+
+//! Test DQN in Cart Pole task.
+BOOST_AUTO_TEST_CASE(CartPoleWithDQN)
+{
+  arma::arma_rng::set_seed(0);
+
+  // Set up the network.
+  FFN<MeanSquaredError<>, GaussianInitialization> model;
+  model.Add<Linear<>>(4, 128);
+  model.Add<ReLULayer<>>();
+  model.Add<Linear<>>(128, 128);
+  model.Add<ReLULayer<>>();
+  model.Add<Linear<>>(128, 2);
+
+  // Set up the optimizer.
+  StandardSGD<decltype(model)> opt(model, 0.0001, 2);
+
+  // Set up the policy and replay method.
+  GreedyPolicy<CartPole> policy(1.0, 1000, 0.1);
+  RandomReplay<CartPole> replayMethod(10, 10000);
+
+  // Set up DQN agent.
+  QLearning<CartPole, decltype(model), decltype(opt), decltype(policy)>
+          agent(model, opt, 0.9, policy, replayMethod, 100, 100);
+
+  arma::running_stat<double> averageReturn;
+  while (true)
+  {
+    double episodeReturn = agent.Episode();
+    averageReturn(episodeReturn);
+    /**
+     * Reaching running average return 60 is enough to show it works.
+     * For the speed of the test case, I didn't set high criterion.
+     */
+    Log::Debug << "Average return: " << averageReturn.mean() << " Episode return: "
+        << episodeReturn << std::endl;
+    if (averageReturn.mean() > 60)
+    {
+      agent.Deterministic() = true;
+      arma::running_stat<double> testReturn;
+      for (size_t i = 0; i < 30; ++i)
+        testReturn(agent.Episode());
+      Log::Debug << "Average return in deterministic test: " << testReturn.mean()
+          << std::endl;
+      break;
+    }
+  }
+}
+
+//! Test Double DQN in Cart Pole task.
+BOOST_AUTO_TEST_CASE(CartPoleWithDoubleDQN)
+{
+  arma::arma_rng::set_seed(0);
```

Comment:
Just for further clarification---if you set a static random seed, you may be lucky and the code may work for that specific seed only.

If you set a seed like `math::RandomSeed(std::time(NULL))` (which will set both random seed for Armadillo and mlpack RNGs, so that is probably the better way to do it), then it will be different on every single run, making debugging difficult.

If instead you leave the random seed completely unset across all tests, then effectively the random seed is set at compile time, meaning that across systems (or across the matrix build) you'll get a good picture of whether or not the test is robust across random seeds, and if it does fail, you can put a debugger on the program without recompiling with a specific random seed in order to see what's wrong.

Typically, before I check in a test that has randomness that I'm not confident of, I'll use `math::RandomSeed(std::time(NULL))` at the start of the test, then run many times:

```
while(true); do bin/mlpack_test -t TestSuiteName/TestName; sleep 1; done
```

(or sometimes we get it wrong and we only find out through the matrix build that some of the tests are failing intermittently, that's ok too as long as we fix it eventually...)